rectools/metrics/popularity.py

```diff
@@ -0,0 +1,160 @@
+    for this item.
+
+    .. math::
+        ARP@k = \frac{1}{\left|U_{t}\right|}\sum_{u\in U_{t}^{}}\frac{\sum_{i\in L_{u}}\phi (i)}{\left| L_{u} \right |}
+
+    where
+    :math:`\phi (i)` is the number of times item i has been rated in the training set.
+    :math:`|U_{t}|` is the number of users in the test set.
+    :math:`L_{u}` is the list of recommended items for user u.
+
+    Parameters
+    ----------
+    k : int
+        Number of items at the top of recommendations list that will be used to calculate metric.
+
+    Examples
+    --------
+    >>> reco = pd.DataFrame(
+    ...     {
+    ...         Columns.User: [1, 1, 2, 2, 2, 3, 3],
+    ...         Columns.Item: [1, 2, 3, 1, 2, 3, 2],
+    ...         Columns.Rank: [1, 2, 1, 2, 3, 1, 2],
+    ...     }
+    ... )
+    >>> prev_interactions = pd.DataFrame(
+    ...     {
+    ...         Columns.User: [1, 1, 2, 2, 3, 3],
+    ...         Columns.Item: [1, 2, 1, 3, 1, 2],
+    ...     }
+    ... )
+    >>> AvgRecPopularity(k=1).calc_per_user(reco, prev_interactions).values
+    array([3., 1., 1.])
+    >>> AvgRecPopularity(k=3).calc_per_user(reco, prev_interactions).values
+    array([2.5, 2. , 1.5])
+    """
+
+    def calc(self, reco: pd.DataFrame, prev_interactions: pd.DataFrame) -> float:
+        """
+        Calculate metric value.
+
+        Parameters
+        ----------
+        reco : pd.DataFrame
+            Recommendations table with columns `Columns.User`, `Columns.Item`, `Columns.Rank`.
+        prev_interactions : pd.DataFrame
+            Table with previous user-item interactions,
+            with columns `Columns.User`, `Columns.Item`.
+
+        Returns
+        -------
+        float
+            Value of metric (average between users).
+        """
+        per_user = self.calc_per_user(reco, prev_interactions)
+        return per_user.mean()
+
+    def calc_per_user(
+        self,
+        reco: pd.DataFrame,
+        prev_interactions: pd.DataFrame,
+    ) -> pd.Series:
+        """
+        Calculate metric values for all users.
+
+        Parameters
+        ----------
+        reco : pd.DataFrame
+            Recommendations table with columns `Columns.User`, `Columns.Item`, `Columns.Rank`.
+        prev_interactions : pd.DataFrame
+            Table with previous user-item interactions,
+            with columns `Columns.User`, `Columns.Item`.
+
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        item_popularity = prev_interactions[Columns.Item].value_counts()
+        item_popularity.name = "popularity"
+
+        reco_k = reco.query(f"{Columns.Rank} <= @self.k")
+        reco_max_k = reco_k.groupby(Columns.User)[Columns.Rank].transform("count")
+        reco_prepared = reco_k.join(item_popularity, on=Columns.Item, how="left")
+        reco_prepared["popularity"] = reco_prepared["popularity"].fillna(0) / reco_max_k
```

Comment:
`pandas` docs don't say that order after `join` is preserved, so we can't rely on it

It's better to add a column with the number of reco to this dataframe

Reply:
Btw, it's more efficient to do actions in the next order:
1. join reco to `item_popularity`
2. `fillna`
3. group by user and calculate simultaneously 1) sum of popularity, 2) number of reco
4. divide 

It's more efficient because you don't need to store `reco_max_k` for long dataframe with useless repeated values, and also because you execute division only for small grouped data

Reply:
Calculations logic was changed to `agg`. It's a bit more time consuming, but no additional memory is used and code looks uncluttered. I guess it's a fair trade-off here.

Reply:
Let me disagree with this

It's much more time consuming, and additional memory consuming is really small

<img width="1025" alt="image" src="https://github.com/MobileTeleSystems/RecTools/assets/14802567/29a9e320-00fc-4a7b-8ee8-63000fd33aa3">

Reply:
And, btw, `sum / count` is the same as `mean` :)

Reply:
Fair enough :)